File: server/csharp/MetaObjects/AgentContext/Stack.cs
// Stack — the resolved tech-stack of a consumer project, the input to the
// agent-context assembler.
//
// Cross-port contract (must match the TS / Java / Python references exactly):
//   - Servers — deduped, in SERVER_LANGS order.
//   - Clients — deduped, in CLIENT_FRAMEWORKS order.
//   - Tokens  — Servers ∪ Clients ∪ {"migration"}, the install-selection set
//     used to choose which reference fragments to emit.
//
// Use Stack.Of(...) to build one: it dedupes + canonical-orders the inputs and
// derives the token set. The canonical orderings are the allow-list — unknown
// entries are dropped — so the result is exactly SERVER_LANGS / CLIENT_FRAMEWORKS
// filtered to the requested set (matching the TS/Java/Python makeStack).

namespace MetaObjects.AgentContext;

/// <summary>
/// The resolved tech-stack of a consumer project — input to the agent-context assembler.
/// </summary>
public sealed class Stack
{
    /// <summary>Server languages, in canonical dedupe order.</summary>
    public static readonly IReadOnlyList<string> ServerLangs =
        new[] { "typescript", "java", "kotlin", "csharp", "python" };

    /// <summary>Client frameworks, in canonical dedupe order.</summary>
    public static readonly IReadOnlyList<string> ClientFrameworks =
        new[] { "react", "tanstack", "angular" };

    /// <summary>Always-present token: schema migrations are TS-owned for every port (ADR-0015).</summary>
    public const string MigrationToken = "migration";

    /// <summary>The five skills, in the exact emit order (matches the TS/Java/Python references).</summary>
    public static readonly IReadOnlyList<string> SkillNames = new[]
    {
        "metaobjects-authoring",
        "metaobjects-codegen",
        "metaobjects-runtime-ui",
        "metaobjects-prompts",
        "metaobjects-verify",
    };

    /// <summary>Deduped servers, in <see cref="ServerLangs"/> order.</summary>
    public IReadOnlyList<string> Servers { get; }

    /// <summary>Deduped clients, in <see cref="ClientFrameworks"/> order.</summary>
    public IReadOnlyList<string> Clients { get; }

    /// <summary><c>servers ∪ clients ∪ {"migration"}</c> — the install-selection set.</summary>
    public IReadOnlySet<string> Tokens { get; }

    private Stack(IReadOnlyList<string> servers, IReadOnlyList<string> clients, IReadOnlySet<string> tokens)
    {
        Servers = servers;
        Clients = clients;
        Tokens = tokens;
    }

    /// <summary>
    /// Build a <see cref="Stack"/>: dedupe + canonical-order the inputs, derive tokens.
    /// Unknown entries are dropped (the canonical orderings are the allow-list).
    /// </summary>
    public static Stack Of(IEnumerable<string> servers, IEnumerable<string> clients)
    {
        var serverSet = new HashSet<string>(servers);
        var clientSet = new HashSet<string>(clients);
        var s = ServerLangs.Where(serverSet.Contains).ToArray();
        var c = ClientFrameworks.Where(clientSet.Contains).ToArray();
        var tokens = new HashSet<string>(s);
        foreach (var x in c) tokens.Add(x);
        tokens.Add(MigrationToken);
        return new Stack(s, c, tokens);
    }
}

File: server/java/metadata/src/main/java/com/metaobjects/agentcontext/Stack.java
package com.metaobjects.agentcontext;

import java.util.List;
import java.util.Set;

/**
 * The resolved tech-stack of a consumer project — the input to the agent-context
 * assembler.
 *
 * <p>Cross-port contract (must match the TypeScript / Python references exactly):
 * <ul>
 *   <li>{@code servers} — deduped, in {@link #SERVER_LANGS} order.</li>
 *   <li>{@code clients} — deduped, in {@link #CLIENT_FRAMEWORKS} order.</li>
 *   <li>{@code tokens} — {@code servers ∪ clients ∪ {"migration"}}, the
 *       install-selection set used to choose which reference fragments to emit.</li>
 * </ul>
 *
 * <p>Use {@link #of(List, List)} to build one: it dedupes + canonical-orders the
 * inputs and derives the token set. The canonical orderings are the allow-list —
 * unknown entries are dropped — so the result is exactly {@link #SERVER_LANGS} /
 * {@link #CLIENT_FRAMEWORKS} filtered to the requested set (matching the TS/Python
 * {@code makeStack}).
 */
public final class Stack {

    /** Server languages, in canonical dedupe order. */
    public static final List<String> SERVER_LANGS =
            List.of("typescript", "java", "kotlin", "csharp", "python");

    /** Client frameworks, in canonical dedupe order. */
    public static final List<String> CLIENT_FRAMEWORKS =
            List.of("react", "tanstack", "angular");

    /** Always-present token: schema migrations are TS-owned for every port (ADR-0015). */
    public static final String MIGRATION_TOKEN = "migration";

    /** The five skills, in the exact emit order (matches the TS/Python references). */
    public static final List<String> SKILL_NAMES = List.of(
            "metaobjects-authoring",
            "metaobjects-codegen",
            "metaobjects-runtime-ui",
            "metaobjects-prompts",
            "metaobjects-verify");

    private final List<String> servers;
    private final List<String> clients;
    private final Set<String> tokens;

    private Stack(List<String> servers, List<String> clients, Set<String> tokens) {
        this.servers = servers;
        this.clients = clients;
        this.tokens = tokens;
    }

    /**
     * Build a {@link Stack}: dedupe + canonical-order the inputs, derive tokens.
     * Unknown entries are dropped (the canonical orderings are the allow-list).
     */
    public static Stack of(List<String> servers, List<String> clients) {
        List<String> s = SERVER_LANGS.stream().filter(servers::contains).toList();
        List<String> c = CLIENT_FRAMEWORKS.stream().filter(clients::contains).toList();
        java.util.LinkedHashSet<String> t = new java.util.LinkedHashSet<>();
        t.addAll(s);
        t.addAll(c);
        t.add(MIGRATION_TOKEN);
        return new Stack(s, c, Set.copyOf(t));
    }

    /** Deduped servers, in {@link #SERVER_LANGS} order. */
    public List<String> servers() {
        return servers;
    }

    /** Deduped clients, in {@link #CLIENT_FRAMEWORKS} order. */
    public List<String> clients() {
        return clients;
    }

    /** {@code servers ∪ clients ∪ {"migration"}} — the install-selection set. */
    public Set<String> tokens() {
        return tokens;
    }
}
